### PR TITLE
add support for ignoring fields

### PR DESCRIPTION
**Pull request by grossvater**

### CODE REVIEW

**Comment by beluchin** [comment]
Many thanks for giving this lib a chance! I hope you are finding it useful. 

Regarding your PR, did you notice you can override field comparators? That is, you can override a field comparator to ignore fields as in:
```java
@Test
public void ignoreFields() {
    class Foo {
        private int great;

        public Foo(int great) {
            this.great = great;
        }

        public int zero() { return 0; }
        public int great() { return great; }
    }

    assertTrue(withOptions()
            .override(comparator(field(Foo.class, "great"), (x, y) -> true))
            .deepEquals(Foo.class, new Foo(1), new Foo(2)));
}
```

**Comment by grossvater** [comment]
Yes, I saw the feature, but it doesn't work with null fields, am I wrong?

**Comment by beluchin** [comment]
what do you mean by _null_ fields? Could you post a test that shows the issue?

**Comment by grossvater** [comment]
Hmmm..... my original try was:

```
    @Test
    public void ignoreFields() {
    	class Bar {
    	}
    	
        class Foo {
            private Bar zero;

            public Foo(Bar zero) {
                this.zero = zero;
            }

            public Bar zero() { return this.zero; }
        }

        assertTrue(withOptions()
                .override(comparator(field(Bar.class, "zero"), (x, y) -> true))
                .deepEquals(Foo.class, new Foo(new Bar()), new Foo(null)));
    }   
```

which fails, because (I see now) the override line is wrong. Instead it should be: 
``` 
   .override(comparator(field(Foo.class, "zero"), (x, y) -> true))
```

My bad.

**Comment by grossvater** [comment]
So ignoring fields feature becomes useless now I think. The only scenario where it might be useful is when one of the fields throws an exception... 

```
    @Test
    public void _20_ignore() {
        class Foo {
        	private int great;
        	
        	public Foo() {
        	}
        	
            public int optional() { throw new RuntimeException(); }
        }
        
        assertTrue(withOptions()
                .override(comparator(field(Foo.class, "optional"), (x, y) -> true))
                .ignore(Foo.class, "optional")
                .deepEquals(Foo.class, new Foo(), new Foo()));
    }
```

**Comment by beluchin** [comment]
hmm, interesting observation ... Just thinking out loud: my original intent was to compare [value objects](https://en.wikipedia.org/wiki/Value_object). What would be a use case in which a value object throws an exception when a attempting to get the value of a field?

**Comment by grossvater** [comment]
I don't use the library for value objects only (I try to compare abstract syntax trees, I will see how suitable is -- I like the library for its simplicity). 
Excepting bugs, a legitimate case for throwing an exception might be:

```
throw new RuntimeException("Not implemented / not supported.");
```

**Comment by beluchin** [comment]
Clearly, support for **not** calling specific methods needs to be added. I will review your code and will get back to you if I have questions.

(Btw, I would argue that AST 's are value objects - data-carrying immutable objects without identity)

**Comment by grossvater** [comment]
Thanks. I think I have to add more tests. I saw there are test methods starting with one, two or no underscore at all, would you tell me what's the convention?

**Comment by beluchin** [comment]
the convention is to give the reader a suggested sequence of features to look into.

File: src/test/java/deepequals/DeepEqualsTest.java [hunk 1]
```diff
@@ -480,6 +480,61 @@ public void baz() {}
                 .deepEquals(Foo.class, new Foo(), new Foo()));
     }
 
+    @Test
+    public void _19_ignore() {
```
Comment (beluchin) [inline src/test/java/deepequals/DeepEqualsTest.java:484]: this test is redundant. Test 20 is sufficient to demonstrate the functionality.

File: src/test/java/deepequals/DeepEqualsTest.java [hunk 2]
```diff
@@ -480,6 +480,61 @@ public void baz() {}
...
+            public int optional() { throw new RuntimeException(); }
+        }
+        
+        assertTrue(withOptions()
+                .ignore(Foo.class, "optional")
+                .deepEquals(Foo.class, new Foo(), new Foo()));
+    }
+
+    @Test
+    public void _21_ignoreNullField() {
```
Comment (beluchin) [inline src/test/java/deepequals/DeepEqualsTest.java:519]: this test is also redundant. Test number _20_ covers it all.

File: src/main/java/deepequals/DeepEquals.java [hunk 3]
```diff
@@ -201,6 +203,12 @@ public WithOptions verbose() {
             return this;
         }
 
+        @Override
+        public WithOptions ignore(Class<?> c, final String name) {
```
Comment (beluchin) [inline src/main/java/deepequals/DeepEquals.java:207]: consider supporting generics as well - via `TypeToken`. Here is an additional test we could write to show the support for generics:

```java
@Test
public void ignoreFieldGenerics() {
    class Foo<T> {
        public T t() { throw new RuntimeException(); }
    }
    class Wrapper {
        private final int i;
        public Wrapper(int i) { this.i = i; }
        public Foo<String> fooString() { return new Foo<>(); }
        public Foo<Integer> fooInteger() {
            return new Foo<Integer>() {
                @Override public Integer t() { return i; }
            };
        }
    }

    assertFalse(withOptions()
            .ignore(new TypeToken<Foo<String>>() {}, "t")
            .deepEquals(Wrapper.class, new Wrapper(1), new Wrapper(42)));
}
```

**Review by beluchin — APPROVED** [review]

**Comment by beluchin** [comment]
fyi - this is all in master now, including the support for generics as indicated [above](https://github.com/beluchin/deepequals/pull/3#discussion_r259637498).

@grossvater - many thanks for your assistance. Let me know if you need further clarification on the implementation.

**Comment by grossvater** [comment]
@beluchin I was busy last weeks. Thank you for supporting this feature.

**Pull request merged**